src/load_cell_endstop.c

```diff
@@ -0,0 +1,297 @@
+        trsync_do_trigger(lce->ts, lce->trigger_reason);
+    }
+}
+
+void
+try_trigger_error(struct load_cell_endstop *lce)
+{
+    uint8_t is_homing = is_flag_set(FLAG_IS_HOMING, lce);
+    if (is_homing) {
+        trsync_do_trigger(lce->ts, lce->error_reason);
+    }
+}
+
+// Used by Sensors to report new raw ADC sample
+void
+load_cell_endstop_report_sample(struct load_cell_endstop *lce
+                                , const int32_t sample, const uint32_t ticks)
+{
+    // save new sample
+    lce->last_sample = sample;
+    lce->last_sample_ticks = ticks;
+    lce->watchdog_count = 0;
+    // check for safety limit violations first
+    const uint8_t is_safety_trigger = sample <= lce->safety_counts_min
+                                        || sample >= lce->safety_counts_max;
+    // too much force, this is an error while homing
+    if (is_safety_trigger) {
+        try_trigger_error(lce);
+    }
+
+    // only trigger when trigger_grams is greater than 0
+    if (lce->trigger_grams == 0) {
+        // clear "live" trigger view for QUERY_ENDSTOPS
+        clear_flag(FLAG_IS_TRIGGERED, lce);
+        return;
+    }
+
+    uint8_t is_trigger = 0;
+    const fixedQ48_t raw_grams = counts_to_grams(lce, sample);
+    const uint8_t is_overflow = overflows_int32(raw_grams);
+    if (!is_overflow) {
+        const fixedQ16_t filtered_grams = sosfilt(lce->sf
+            , (fixedQ16_t)raw_grams);
+        is_trigger = abs(filtered_grams) >= lce->trigger_grams_fixed;
+    } else {
+        // assume triggered in the case of an overflow
+        is_trigger = 1;
+        // While homing an overflow is an error
+        try_trigger_error(lce);
+    }
+
+    // latching trigger for QUERY_ENDSTOPS: dont clear until force is removed
+    uint8_t alredy_triggered = is_flag_set(FLAG_IS_TRIGGERED, lce);
+    if (alredy_triggered) {
+        const int64_t abs_int_grams = abs(round_fixedQ48(raw_grams));
+        uint8_t is_abs_trigger = abs_int_grams >= lce->trigger_grams;
+        is_trigger = is_abs_trigger || is_trigger;
+    }
+
+    // update trigger state
+    if (is_trigger) {
+        try_trigger(lce, ticks);
+    }
+    else {
+        // clear "live" trigger view for QUERY_ENDSTOPS
+        clear_flag(FLAG_IS_TRIGGERED, lce);
+    }
+}
+
+// Timer callback that monitors for timeouts
+static uint_fast8_t
+watchdog_event(struct timer *t)
+{
+    struct load_cell_endstop *lce = container_of(t, struct load_cell_endstop
+                                        , time);
+    uint8_t is_homing = is_flag_set(FLAG_IS_HOMING, lce);
+    uint8_t is_homing_trigger = is_flag_set(FLAG_IS_HOMING_TRIGGER, lce);
+    // the watchdog stops when not homing or when trsync becomes triggered
+    if (!is_homing || is_homing_trigger) {
+        return SF_DONE;
+    }
+
+    irq_disable();
+    if (lce->watchdog_count > lce->watchdog_max) {
+        try_trigger_error(lce);
+    }
+    lce->watchdog_count += 1;
+    irq_enable();
```

Comment:
Eeks!  Interrupts are always disabled in a timer, and if you call `irq_enable()` from a timer it can destabilize the system.

Reply:
Fixed, removed irq_disable() / irq_enable()